### ISSUE TITLE
Freeze_graph results in very poor accuracy compared to manually exporting and freezing the graph?

### ISSUE THREAD

**Issue by kwotsin** [tensorflow/tensorflow#9724]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04 LTS
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**: 1.1
- **Bazel version (if compiling from source)**: 4.5
- **CUDA/cuDNN version**: 8.0/5.1
- **GPU model and memory**: GTX 860M

### Describe the problem
From using the official freeze_graph.py file from TF, I am getting a very low accuracy in prediction as compared to manually exporting the graph using a file I wrote called `write_pb.py`, I get a much higher accuracy.

To be specific, here are the differences:

**Differences:**
1. Using `write_pb.py` to manually export the graph converted way more variables to constants, even with the same checkpoint files.
2.  It takes a long, long time for `freeze_graph.py` to actually complete the exporting.
3. Very importantly: I get a very low accuracy from using freeze_graph. Meanwhile, by exporting the graph manually, I get a nearly identical accuracy as if I predicted an image right from the checkpoint without freezing.
4. Manually exporting the graph results in a smaller file size (just 1-2MB of difference).
5. The manually exported graph has a faster inference time than the graph obtained from `freeze_graph.py`.

This is the freeze_graph.py file I got from the TF repo: https://gist.github.com/kwotsin/b9dae8246a30371a1a10690e2fa27cb7

This is the `write_pb` file I wrote:
https://gist.github.com/kwotsin/8e43f5db4815e1f1af37da70d0933d8b

### Source code / logs
Using freeze_graph with this command: `python freeze_graph.py --input_checkpoint=model.ckpt-18279 --input_graph=graph.pbtxt --output_graph=frozen_model_from_freeze_graph.pb --output_node_names=InceptionResnetV2/Logits/Predictions`

I get this output:
```
2017-05-06 22:27:22.967898: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-05-06 22:27:22.968268: I tensorflow/core/common_runtime/gpu/gpu_device.cc:887] Found device 0 with properties: 
name: GeForce GTX 860M
major: 5 minor: 0 memoryClockRate (GHz) 1.0195
pciBusID 0000:01:00.0
Total memory: 3.95GiB
Free memory: 3.34GiB
2017-05-06 22:27:22.968283: I tensorflow/core/common_runtime/gpu/gpu_device.cc:908] DMA: 0 
2017-05-06 22:27:22.968288: I tensorflow/core/common_runtime/gpu/gpu_device.cc:918] 0:   Y 
2017-05-06 22:27:22.968306: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 860M, pci bus id: 0000:01:00.0)
2017-05-06 22:27:33.499491: I tensorflow/compiler/xla/service/platform_util.cc:58] platform CUDA present with 1 visible devices
2017-05-06 22:27:33.499519: I tensorflow/compiler/xla/service/platform_util.cc:58] platform Host present with 8 visible devices
2017-05-06 22:27:33.509659: I tensorflow/compiler/xla/service/service.cc:183] XLA service 0x36ced220 executing computations on platform Host. Devices:
2017-05-06 22:27:33.509678: I tensorflow/compiler/xla/service/service.cc:191]   StreamExecutor device (0): <undefined>, <undefined>
2017-05-06 22:27:33.509827: I tensorflow/compiler/xla/service/platform_util.cc:58] platform CUDA present with 1 visible devices
2017-05-06 22:27:33.509837: I tensorflow/compiler/xla/service/platform_util.cc:58] platform Host present with 8 visible devices
2017-05-06 22:27:33.512265: I tensorflow/compiler/xla/service/service.cc:183] XLA service 0x36ce6210 executing computations on platform CUDA. Devices:
2017-05-06 22:27:33.512279: I tensorflow/compiler/xla/service/service.cc:191]   StreamExecutor device (0): GeForce GTX 860M, Compute Capability 5.0
Converted 490 variables to const ops.
7871 ops in the final graph.
```
Meanwhile, if I manually freeze the graph using `write_pb.py`,

I get this output:
```
2017-05-06 22:39:00.197711: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-05-06 22:39:00.198096: I tensorflow/core/common_runtime/gpu/gpu_device.cc:887] Found device 0 with properties: 
name: GeForce GTX 860M
major: 5 minor: 0 memoryClockRate (GHz) 1.0195
pciBusID 0000:01:00.0
Total memory: 3.95GiB
Free memory: 3.29GiB
2017-05-06 22:39:00.198120: I tensorflow/core/common_runtime/gpu/gpu_device.cc:908] DMA: 0 
2017-05-06 22:39:00.198125: I tensorflow/core/common_runtime/gpu/gpu_device.cc:918] 0:   Y 
2017-05-06 22:39:00.198143: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 860M, pci bus id: 0000:01:00.0)
2017-05-06 22:39:00.709417: I tensorflow/compiler/xla/service/platform_util.cc:58] platform CUDA present with 1 visible devices
2017-05-06 22:39:00.709452: I tensorflow/compiler/xla/service/platform_util.cc:58] platform Host present with 8 visible devices
2017-05-06 22:39:00.710238: I tensorflow/compiler/xla/service/service.cc:183] XLA service 0xef16320 executing computations on platform Host. Devices:
2017-05-06 22:39:00.710252: I tensorflow/compiler/xla/service/service.cc:191]   StreamExecutor device (0): <undefined>, <undefined>
2017-05-06 22:39:00.710374: I tensorflow/compiler/xla/service/platform_util.cc:58] platform CUDA present with 1 visible devices
2017-05-06 22:39:00.710384: I tensorflow/compiler/xla/service/platform_util.cc:58] platform Host present with 8 visible devices
2017-05-06 22:39:00.710645: I tensorflow/compiler/xla/service/service.cc:183] XLA service 0xef18180 executing computations on platform CUDA. Devices:
2017-05-06 22:39:00.710654: I tensorflow/compiler/xla/service/service.cc:191]   StreamExecutor device (0): GeForce GTX 860M, Compute Capability 5.0
Exporting graph...
Converted 898 variables to const ops.
```

**Comment by petewarden** [comment]
It's hard to say without knowing how the graph was exported, but in your script you set `is_training=False` when creating the Slim model. This will create a different model with some training-only variables and ops removed, which can be important for good results when running the graph.

**Comment by xiusir** [comment]
Same.

I use freeze_graph to export my model to a file named "frozen.pb". But Found that the accuracy of predictions on frozen.pb is very poor.

I know the problem maybe MovingAverage not included in frozen.pb or not loaded in prediction.

When I use model.ckpt files to restore model for evaluating, if i call tf.train.ExponentialMovingAverage(0.999) , then the accuracy is good as expected, else the accuracy is bad.

So How To export a binary model which performance is the same as the one restored from checkpoint files? I want to use ".pb" files in Android Devices.

Freeze Command:
```
~/bazel-bin/tensorflow/python/tools/freeze_graph \
  --input_graph=./graph.pbtxt \
  --input_checkpoint=./model.ckpt-100000 \
  --output_graph=frozen.pb \
  --output_node_names=output  \
  --restore_op_name=save/restore_all \
  --clear_devices
```
[More code at StackOverflow](https://stackoverflow.com/questions/44690363/how-to-use-tf-train-exponentialmovingaverage-in-android-ios)

**Comment by xiusir** [comment]
@kwotsin 
Thank you !   Your code is also helpful for me, it works.

**Comment by MyHumbleSelf** [comment]
I have the exact same issue and have been hopelessly confused about this, thinking that the `freeze_graph.py` script couldn't be the breaking point.

In my case, fine-tuning off MobileNet, the accuracy dropped to the abyss and the network outputted form `freeze_graph.py` was extremely degenerate – tiny variations in the input image would lead to different top classes being predicted. I printed and various weights and they matched.

I confirmed @petewarden's suspicion that if a network exported directly (like in @kwotsin's script) works iff is_training=False is passed. I haven't debugged further, but even if this doesn't have much to do with the `freeze_graph.py` script itself, some docs should probably be adjusted.

**Comment by tensorflowbutler** [comment]
This issue is automatically closed due to lack of activity. Please re-open if this is still an issue for you. Thanks!

**Comment by Sixigma** [comment]
Hi , I encounter exactly the same problem here 
But the script I use , which follows the tutorial of Slim , is a little different from kwotsin's script .
The first step is to exporting the inference graph , here is the script :
```
python export_inference_graph.py \
--alsologtostderr \
--model_name=mobilenet_v1 \
--dataset_name=flowers \
--batch_size=1 \
--image_size=224 \
--output_file=.../mobilenet_224/mobilenet_v1.pb
```
And freeze the checkpoint by freeze_graph.py :
```
bazel-bin/tensorflow/python/tools/freeze_graph \
--input_checkpoint=.../model.ckpt-10000 \
--input_graph=.../mobilenet_224/mobilenet_v1.pb \
--input_binary=True \
--output_graph=.../frozen_graph.pb \
--output_node_names=MobilenetV1/Predictions/Reshape_1
```
The validation dataset , which is totally separate from training dataset , is evaluated about 70% accuracy ,  but rapidly decrease after freezing.
What may result in this problem? Is it solved?

**Comment by Karthik777** [comment]
Is there a fix for this issue yet. having the same problem. I've tried @kwotsin's script . It still gives me inaccurate results when loaded with the graph.pb file

**Comment by ambr89** [comment]
Hello!
I've the same problem!
@Karthik777  any news?

**Comment by Aurora11111** [comment]
I meet the same problem ,any solutions?

**Comment by Karthik777** [comment]
@ambr89 : I fixed by using transformations on graph with TransformGraph.
from tensorflow.tools.graph_transforms import TransformGraph
        transforms = ['add_default_attributes',
                      'remove_nodes(op=Identity, op=CheckNumerics)',
                      'fold_batch_norms', 'fold_old_batch_norms',
                      'strip_unused_nodes', 'sort_by_execution_order']
        transformed_graph_def = TransformGraph(tf.get_default_graph().as_graph_def(),'Placeholder', output_node_names.split(","), transforms)

        # We use a built-in TF helper to export variables to constants
        output_graph_def = tf.graph_util.convert_variables_to_constants(
            sess,  # The session is used to retrieve the weights
            transformed_graph_def,  # The graph_def is used to retrieve the nodes
            output_node_names.split(",")  # The output node names are used to select the usefull nodes
        )`
`

**Comment by ktiwary2** [comment]
> @ambr89 : I fixed it by importing meta graph and applying transformations on graph with TransformGraph.
> 
> ```
> input_checkpoint = "chckpoint path"
> with tf.Session(graph=tf.Graph()) as sess:
>         # We import the meta graph in the current default Graph
>         saver = tf.train.import_meta_graph(input_checkpoint + '.meta', clear_devices=clear_devices)
> 
>         # We restore the weights
>         saver.restore(sess, input_checkpoint)
> 
>         from tensorflow.tools.graph_transforms import TransformGraph
>         transforms = ['add_default_attributes',
>                       'remove_nodes(op=Identity, op=CheckNumerics)',
>                       'fold_batch_norms', 'fold_old_batch_norms',
>                       'strip_unused_nodes', 'sort_by_execution_order']
>         transformed_graph_def = TransformGraph(tf.get_default_graph().as_graph_def(),'Placeholder', output_node_names.split(","), transforms)
> 
> # We use a built-in TF helper to export variables to constants
>        output_graph_def = tf.graph_util.convert_variables_to_constants(
>             sess,  # The session is used to retrieve the weights
>             transformed_graph_def,  # The graph_def is used to retrieve the nodes
>             output_node_names.split(",")  # The output node names are used to select the useful nodes
>         )
>       with tf.gfile.GFile("optimised_model.pb", "wb") as f:
>             f.write(output_graph_def.SerializeToString())
> ```
> Change Placeholder to input node name and fill in your output nodes.

Yes, you're right to do that. This particular function literally just converts all the variables into constants without doing any kind of transformations to it. This is a bug I found to mess completely with my accuracy as well. So, it expects a subgarph to be fed in with only those inputs and outputs that are necessary. I suggest using [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py) since this strips the graphs of any Restore Tensor ops and file loading calls.

**Comment by schliffen** [comment]
Thank you @ktiwary2 for your solution. I still have difficulty to find proper output nodes of SSD-Mobilenet when I loaded it from checkpoints. There are lots of nodes there 5000+ which I cant check them one by one. I checked some 1000+ last layers, but could not found. Any idea for this?  -Thankyou

**Comment by schliffen** [comment]
It seems this takes frozen graph as input, I wanted the same thing for meta graph "checkpoint.meta" file.
Can we use for meta graphs?

**Comment by zzk88862** [comment]
> @kwotsin
> Thank you ! Your code is also helpful for me, it works.

hi,  please tell me how to solve this probleme, i tried many solutions but it is not useful

**Comment by zzk88862** [comment]
> @kwotsin
> Thank you ! Your code is also helpful for me, it works.



> Same Problem.
> @petewarden
> 
> I use freeze_graph to export my model to a file named "frozen.pb". But Found that the accuracy of predictions on frozen.pb is very poor.
> 
> I know the problem maybe MovingAverage not included in frozen.pb or not loaded in prediction.
> 
> When I use model.ckpt files to restore model for evaluating, if i call tf.train.ExponentialMovingAverage(0.999); variables_to_restore = variable_averages.variables_to_restore(); saver = tf.train.Saver(variables_to_restore) , then the accuracy is good as expected, else the accuracy is bad.
> 
> So How To export a binary model which performance is the same as the one restored from checkpoint files? I want to use ".pb" files in Android Devices.
> 
> Thank you!
> 
> Freeze Command:
> 
> ```
> ~/bazel-bin/tensorflow/python/tools/freeze_graph \
>   --input_graph=./graph.pbtxt \
>   --input_checkpoint=./model.ckpt-100000 \
>   --output_graph=frozen.pb \
>   --output_node_names=output  \
>   --restore_op_name=save/restore_all \
>   --clear_devices
> ```
> [More code at StackOverflow](https://stackoverflow.com/questions/44690363/how-to-use-tf-train-exponentialmovingaverage-in-android-ios)
hi,  can you  tell me how to solve this problem, i tried many solutions but it si not useful
